### PR TITLE
Sessionless - kotlin

### PR DESCRIPTION
**Pull request by Galacticai**
# Sessionless in kotlin
### Module:
src/kotlin/standalone/kotlin/
### Example (Spring boot web mvc)
src/kotlin/example/spring/

# Highlights:
- ### util/
    Contains utilities for generating and converting keys... and more 
- ### models/
    Interfaces or data classes that are meant to represent the data structures or general skeletons used in `impl/...`
- ### impl/
    Implementations of `models/`

# Main focus:
- ### ..../models/**ISessionless.kt**
- ### ..../impl/**Sessionless.kt** (implementation of `ISessionless`)
   - ### `Sessionless.WithKeyStore`
       Sessionless implementation with native support for `java.security.KeyStore`
   - ### `Sessionless.WithCustomVault`
       Sessionless implementation while allowing custom storage management (`ICustomVault`, with custom `get`/`save` functions)

# Notice:
Almost all classes, objects, and functions are documented using comment docs which means that your IDE will show you the documentation while typing/hovering/etc...

### CODE REVIEW

**Comment by Galacticai** [comment]
# [ISessionless](./src/main/kotlin/com/planetnine/sessionless/models/ISessionless.kt):

```mermaid
classDiagram
  class ISessionless {
    +vault: IVault

    +generateUUID(): String
    +sign(message: String, key: PrivateKey): IMessageSignature
    +verify(signedMessage: SignedMessage): Boolean
    +associate(...signedMessages: SignedMessage): Boolean
  }

  class IWithKeyStore {
    @Override
    ^vault: IKeyStoreVault

    +generateKeys(access: KeyAccessInfo): KeyPair
    +generateKeysAsync(access: KeyAccessInfo): KeyPair
    +getKeys(): KeyPair
    +sign(message: String, access: KeyAccess): IMessageSignature
  }

  class IWithCustomVault {
    @Override
    ^vault: ICustomVault

    +generateKeys(): KeyPairHex
    +generateKeysAsync(): KeyPairHex
    +getKeys(): KeyPairHex
    +sign(message: String): SignedMessage
  }

  ISessionless <-- IWithKeyStore
  ISessionless <-- IWithCustomVault
```

# [MessageSignature](./src/main/kotlin/com/planetnine/sessionless/impl/MessageSignature.kt):

```mermaid
classDiagram
  class IMessageSignature { }
  class MessageSignatureInt {
    +r: BigInteger
    +s: BigInteger
    +constructor(r: BigInteger, s: BigInteger)
    +toHex(): MessageSignatureHex
  }
  class MessageSignatureHex {
    +rHex: String
    +sHex: String
    +constructor(rHex: String, sHex: String)
    +toInt(): MessageSignatureInt
  }

  IMessageSignature <-- MessageSignatureInt
  IMessageSignature <-- MessageSignatureHex
```

# [IVault](./src/main/kotlin/com/planetnine/sessionless/models/IVault.kt):

```mermaid
classDiagram
  class IVault { }
  class IKeyStoreVault {
    +keyStore: KeyStore

    +save(pair: KeyPair, access: KeyAccessInfo, certificate: Certificate)
    +get(access: KeyAccessInfo): KeyPair
  }
  class ICustomVault {
    +save(pair: SimpleKeyPair)
    +get(): SimpleKeyPair?
  }

  class KeyStoreVault {
    +save(pair: KeyPair, access: KeyAccessInfo)
    +constructor(keyStore: KeyStore)
  }

  IVault <-- IKeyStoreVault
  IVault <-- ICustomVault
  IKeyStoreVault <-- KeyStoreVault
```

# The rest:

```mermaid
classDiagram
  class SignedMessage {
    +message: String
    +signature: IMessageSignatureHex
    +publicKey: String
    +constructor(..all..)
  }

  class KeyPairHex {
    +privateKey: String
    +publicKey: String
    +constructor(..all..)
  }

  class KeyAccessInfo {
    +alias: String
    +password: CharArray?
    +constructor(..all..)
    +constructor(alias: String, password: String)
  }
```

File: src/kotlin/example/spring/src/main/kotlin/com/planetnine/sessionless/controllers/SessionlessController.kt [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+
+    data class CoolStuffResponse(val doubleCool: String?, val error: String?)
+    data class CoolStuffReqBody(
+        val userUUID: String, val signature: MessageSignature?,
+        val timestamp: Long, val coolness: String,
+    )
+
+    @PutMapping("/cool-stuff")
+    fun coolStuff(req: HttpServletRequest, @RequestBody body: CoolStuffReqBody): CoolStuffResponse {
+        val publicKey = getUserPublicKey(body.userUUID)
```
Comment (zach-planet-nine) [inline src/kotlin/example/spring/src/main/kotlin/com/planetnine/sessionless/controllers/SessionlessController.kt:142]: So we standardized on `uuid` for the request bodies (and `pubKey`). I think it's probably worth it to just find and replace all the `userUUID`s with `uuid`, but if you want to keep `userUUID` on internal stuff that's ok, but anything that clients send will be just `uuid`.

File: src/kotlin/example/spring/src/main/kotlin/com/planetnine/sessionless/util/sessionless/models/ISessionless.kt [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+     * @param message The message to be signed.
+     * @return Signature as a [MessageSignature]. */
+    fun sign(message: String, privateKey: PrivateKey): MessageSignature
+
+    /** Verifies a given signature with a public key.
+     * @param identifiableMessage message that was signed
+     * @return True if the [IdentifiableMessage.signature] is valid
+     * for the given [IdentifiableMessage.message] and [IdentifiableMessage.publicKey].
+     * @see sign */
+    fun verify(identifiableMessage: IdentifiableMessage): Boolean
```
Comment (zach-planet-nine) [inline src/kotlin/example/spring/src/main/kotlin/com/planetnine/sessionless/util/sessionless/models/ISessionless.kt:69]: This should be `verifySignature` to stay consistent with the other implementations

File: src/kotlin/standalone/sessionless/src/main/kotlin/com/planetnine/sessionless/impl/Sessionless.kt [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+            messageHash,
+            signature.r,
+            signature.s
+        )
+    }
+
+    /** Verifies a given signature with a public key
+     * - This calls [verify] with [SignedMessage] out of the provided parameters
+     * @see verify */
+    fun verify(message: String, publicKey: String, signature: MessageSignatureHex): Boolean {
```
Comment (zach-planet-nine) [inline src/kotlin/standalone/sessionless/src/main/kotlin/com/planetnine/sessionless/impl/Sessionless.kt:124]: same here with `verifySignature`

**Review by zach-planet-nine — APPROVED** [review]
Looks great! Just some naming things to update for consistency

File: src/kotlin/example/spring/src/main/kotlin/com/planetnine/sessionless/util/sessionless/models/ISessionless.kt (hunk 2 above)
Comment (Galacticai) [reply src/kotlin/example/spring/src/main/kotlin/com/planetnine/sessionless/util/sessionless/models/ISessionless.kt:69]: Sure thing

I also noticed an error here 
`IdentifiableMessage` was renamed to `SignedMessage` but this file wasn't updated for some reason

**Pull request merged**